### PR TITLE
Add configurable version number to software and index files

### PR DESCRIPTION
**Pull request by mielvds**
This PR implements a more manageable versioning approach to prevent compatibility issues with HDT or index files.
- all tools now print a `vX.Y.Z` version number when using the `-V` (capital) cmd parameter
- index files are generated with a `.vXY` suffix
- HDT file compatibility is checked and ensured with matching `X`
- index file compatibility is ensured with matching `X` and `Y`
- Backwards compatibility is ensured between matching `X` and `Y`, but increasing `Z`.

### CODE REVIEW

File: hdt-lib/include/HDTVocabulary.hpp [hunk 1]
```diff
@@ -33,12 +33,13 @@
...
 
 #include <string>
+#include <HDTVersion.hpp>
 
 namespace hdt {
 namespace HDTVocabulary {
 	// Base
 	const std::string HDT_BASE = "<http://purl.org/HDT/hdt#";
-	const std::string HDT_CONTAINER = HDT_BASE+"HDTv1>";
+	const std::string HDT_CONTAINER = HDT_BASE+"HDTv"+HDT_VERSION+">";
```
Comment (RubenVerborgh) [inline hdt-lib/include/HDTVocabulary.hpp]: Would it make sense to include index and release in here? (Even though only the main version is supposed to induce breaking changes, certain releases might introduce bugs in generated HDT files, as we had before. Might be good to be able to detect them.)

File: hdt-lib/src/hdt/BasicHDT.cpp [hunk 2]
```diff
@@ -656,7 +657,7 @@ void BasicHDT::loadFromHDT(std::istream & input, ProgressListener *listener)
 	controlInformation.load(input);
 	std::string hdtFormat = controlInformation.getFormat();
 	if(hdtFormat!=HDTVocabulary::HDT_CONTAINER) {
```
Comment (RubenVerborgh) [inline hdt-lib/src/hdt/BasicHDT.cpp]: But, as per [my above comment](https://github.com/rdfhdt/hdt-cpp/pull/36/files#r69955509), this check would need to change then.

File: hdt-lib/src/hdt/BasicHDT.cpp [hunk 3]
```diff
@@ -742,7 +743,7 @@ size_t BasicHDT::loadMMap(unsigned char *ptr, unsigned char *ptrMax, ProgressLis
     count+=controlInformation.load(&ptr[count], ptrMax);
     std::string hdtFormat = controlInformation.getFormat();
     if(hdtFormat!=HDTVocabulary::HDT_CONTAINER) {
-    	throw std::runtime_error("This software cannot open this version of HDT File.");
+    	throw std::runtime_error("This software cannot open this version of HDT File (" + hdtFormat + " =/= " + HDTVocabulary::HDT_CONTAINER + ")");
```
Comment (RubenVerborgh) [inline hdt-lib/src/hdt/BasicHDT.cpp]: Idem.

File: hdt-lib/src/hdt/BasicHDT.cpp [hunk 4]
```diff
@@ -777,7 +778,9 @@ size_t BasicHDT::loadMMapIndex(ProgressListener *listener) {
 
     // Get path
     string indexFile(fileName);
-    indexFile.append(".index");
+    indexFile.append(".index.v");
+		indexFile.append(HDT_VERSION);
+		indexFile.append(INDEX_VERSION);
```
Comment (RubenVerborgh) [inline hdt-lib/src/hdt/BasicHDT.cpp]: Indenting seems off.

File: hdt-lib/src/hdt/BasicHDT.cpp [hunk 5]
```diff
@@ -835,7 +838,7 @@ void BasicHDT::saveToHDT(std::ostream & output, ProgressListener *listener)
 
 void BasicHDT::loadOrCreateIndex(ProgressListener *listener) {
 
-	string indexname = this->fileName + ".index";
+	string indexname = this->fileName + ".index.v" + HDT_VERSION + INDEX_VERSION;
```
Comment (RubenVerborgh) [inline hdt-lib/src/hdt/BasicHDT.cpp]: If this concatenation occurs frequently, might be good to have a function / macro for it.

File: hdt-lib/tools/hdt2rdf.cpp [hunk 6]
```diff
@@ -53,12 +55,16 @@ int main(int argc, char **argv) {
...
-	while( (c = getopt(argc,argv,"f:"))!=-1) {
+	while( (c = getopt(argc,argv,"f:V:"))!=-1) {
 		switch(c) {
 		case 'f':
 			rdfFormat = optarg;
 			cout << "Format: " << rdfFormat << endl;
 			break;
+		case 'V':
+			cout << "v" << HDT_VERSION << "." << INDEX_VERSION << "." << RELEASE_VERSION << endl;
+			return 1;
```
Comment (RubenVerborgh) [inline hdt-lib/tools/hdt2rdf.cpp]: Not an error condition; should be `0`.

File: hdt-lib/tools/hdtInfo.cpp [hunk 7]
```diff
@@ -66,6 +67,10 @@ int main(int argc, char **argv) {
 		case 'o':
 			outputFile = optarg;
 			break;
+		case 'V':
+			cout << "v" << HDT_VERSION << "." << INDEX_VERSION << "." << RELEASE_VERSION << endl;
+			return 1;
```
Comment (RubenVerborgh) [inline hdt-lib/tools/hdtInfo.cpp]: Should be `0`.

File: hdt-lib/tools/hdtSearch.cpp [hunk 8]
```diff
@@ -124,6 +125,10 @@ int main(int argc, char **argv) {
 		case 'm':
 			measure = true;
 			break;
+		case 'V':
+			cout << "v" << HDT_VERSION << "." << INDEX_VERSION << "." << RELEASE_VERSION << endl;
+			return 1;
```
Comment (RubenVerborgh) [inline hdt-lib/tools/hdtSearch.cpp]: Should be `0`.

File: hdt-lib/tools/rdf2hdt.cpp [hunk 9]
```diff
@@ -90,6 +92,10 @@ int main(int argc, char **argv) {
 		case 'i':
 			generateIndex=true;
 			break;
+		case 'V':
+			cout << "v" << HDT_VERSION << "." << INDEX_VERSION << "." << RELEASE_VERSION << endl;
+			return 1;
```
Comment (RubenVerborgh) [inline hdt-lib/tools/rdf2hdt.cpp]: Should be `0`.

**Comment by RubenVerborgh** [comment]
Went through the pull request; looks good for me except for the comments I left.
Did not do a global search for `.index`—did you?

File: hdt-lib/include/HDTVocabulary.hpp (hunk 1 above)
Comment (mielvds) [inline hdt-lib/include/HDTVocabulary.hpp]: Hmm, I didn't do that for two reasons: the breaking changes, and I'd had to write a function that parses this URI to compare, but didn't know where to put it.
But what you're saying is also my preference, just more work.

File: hdt-lib/src/hdt/BasicHDT.cpp (hunk 2 above)
Comment (mielvds) [inline hdt-lib/src/hdt/BasicHDT.cpp]: Jup indeed, I sort of started on a parse function: https://github.com/rdfhdt/hdt-cpp/blob/61288cbfb1d0fb3424392d26ea6d20c77d5b8024/hdt-lib/include/HDTVersion.hpp#L9

File: hdt-lib/src/hdt/BasicHDT.cpp (hunk 4 above)
Comment (mielvds) [inline hdt-lib/src/hdt/BasicHDT.cpp]: actually, no :) C++ magic

File: hdt-lib/src/hdt/BasicHDT.cpp [hunk 10]
```diff
@@ -655,8 +656,9 @@ void BasicHDT::loadFromHDT(std::istream & input, ProgressListener *listener)
 	// Load Global ControlInformation.
 	controlInformation.load(input);
 	std::string hdtFormat = controlInformation.getFormat();
-	if(hdtFormat!=HDTVocabulary::HDT_CONTAINER) {
-		throw std::runtime_error("This software cannot open this version of HDT File.");
+	//if(hdtFormat!=HDTVocabulary::HDT_CONTAINER) {
+	if (hdtFormat.find(HDTVocabulary::HDT_CONTAINER_BASE) == std::string::npos) {
```
Comment (mielvds) [inline hdt-lib/src/hdt/BasicHDT.cpp]: @RubenVerborgh this should work

File: hdt-lib/src/hdt/BasicHDT.cpp [hunk 11]
```diff
@@ -655,8 +656,9 @@ void BasicHDT::loadFromHDT(std::istream & input, ProgressListener *listener)
 	// Load Global ControlInformation.
 	controlInformation.load(input);
 	std::string hdtFormat = controlInformation.getFormat();
-	if(hdtFormat!=HDTVocabulary::HDT_CONTAINER) {
-		throw std::runtime_error("This software cannot open this version of HDT File.");
+  //if(hdtFormat!=HDTVocabulary::HDT_CONTAINER) {
+  if (hdtFormat.find(HDTVocabulary::HDT_CONTAINER_BASE) == std::string::npos) {
```
Comment (mielvds) [inline hdt-lib/src/hdt/BasicHDT.cpp]: @RubenVerborgh this should do it.

File: hdt-lib/src/hdt/BasicHDT.cpp [hunk 12]
```diff
@@ -776,8 +779,7 @@ size_t BasicHDT::loadMMapIndex(ProgressListener *listener) {
     }
 
     // Get path
-    string indexFile(fileName);
-    indexFile.append(".index");
+    string indexFile(fileName + ".index.v" + HDT_VERSION + INDEX_VERSION);
```
Comment (mielvds) [inline hdt-lib/src/hdt/BasicHDT.cpp]: got rid of the append

**Comment by mielvds** [comment]
@RubenVerborgh last commits should fix your comments. I did do a search for `.index`, everything should be covered.

File: hdt-lib/src/hdt/BasicHDT.cpp (hunk 2 above)
Comment (RubenVerborgh) [inline hdt-lib/src/hdt/BasicHDT.cpp]: Isn't there any built-in regex support we can rely on (probably not)? Or `scanf`?

File: hdt-lib/src/hdt/BasicHDT.cpp (hunk 4 above)
Comment (RubenVerborgh) [inline hdt-lib/src/hdt/BasicHDT.cpp]: Not even [this one](http://www.cplusplus.com/reference/string/string/operator+/)?

Comment (mielvds) [inline hdt-lib/src/hdt/BasicHDT.cpp]: This has been solved now

File: hdt-lib/src/hdt/BasicHDT.cpp (hunk 2 above)
Comment (mielvds) [inline hdt-lib/src/hdt/BasicHDT.cpp]: I used find, not ok?

Comment (RubenVerborgh) [inline hdt-lib/src/hdt/BasicHDT.cpp]: OK, but the function seems long for the task.

Comment (mielvds) [inline hdt-lib/src/hdt/BasicHDT.cpp]: I think we're having an outdated discussion: https://github.com/rdfhdt/hdt-cpp/pull/36/files#diff-9b14bf8cdb754874c17bea746d616ae3R660

**Comment by bendiken** [comment]
Thanks, will review

**Comment by webdata** [comment]
I have fixed some typo and I think this is ready to merge. I am now preparing the same numbering for hdt-java

**Pull request merged**